go/vt/vtgate/engine/merge.go

```diff
@@ -0,0 +1,72 @@
+package engine
+
+import (
+	"sync"
+
+	"github.com/youtube/vitess/go/sqltypes"
+	"github.com/youtube/vitess/go/vt/vtgate/queryinfo"
+)
+
+// Merge specifies the parameters for a merge primitive
+type Merge struct {
+	Left, Right Primitive `json:",omitempty"`
+}
+
+// Execute performs an execute
+func (merge *Merge) Execute(vcursor VCursor, queryConstruct *queryinfo.QueryConstruct, joinvars map[string]interface{}, wantfields bool) (*sqltypes.Result, error) {
+	var lres, rres *sqltypes.Result
+	var lerr, rerr error
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func(vc *VCursor, qc *queryinfo.QueryConstruct, jv map[string]interface{}, wf bool) {
+		defer wg.Done()
+		lres, lerr = merge.Left.Execute(*vc, qc, jv, wf)
+	}(&vcursor, queryConstruct, joinvars, wantfields)
+	wg.Add(1)
+	go func(vc *VCursor, qc *queryinfo.QueryConstruct, jv map[string]interface{}, wf bool) {
+		defer wg.Done()
+		rres, rerr = merge.Right.Execute(*vc, qc, jv, wf)
+	}(&vcursor, queryConstruct, joinvars, wantfields)
+	wg.Wait()
+	// TODO(acharis): possibly better to aggregate errors as in ScatterConn
+	if lerr != nil {
+		return nil, lerr
+	}
+	if rerr != nil {
+		return nil, rerr
+	}
+	// TODO(acharis): deal with column type mismatch
+	lres.AppendResult(rres)
+	return lres, nil
+}
+
+// StreamExecute performs a streaming execute
+func (merge *Merge) StreamExecute(vcursor VCursor, queryConstruct *queryinfo.QueryConstruct, joinvars map[string]interface{}, wantfields bool, callback func(*sqltypes.Result) error) error {
+	var lerr, rerr error
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func(vc *VCursor, qc *queryinfo.QueryConstruct, jv map[string]interface{}, wf bool, cb func(*sqltypes.Result) error) {
+		defer wg.Done()
+		lerr = merge.Left.StreamExecute(*vc, qc, jv, wf, cb)
+	}(&vcursor, queryConstruct, joinvars, wantfields, callback)
+	wg.Add(1)
+	go func(vc *VCursor, qc *queryinfo.QueryConstruct, jv map[string]interface{}, wf bool, cb func(*sqltypes.Result) error) {
+		defer wg.Done()
+		rerr = merge.Right.StreamExecute(*vc, qc, jv, wf, cb)
+	}(&vcursor, queryConstruct, joinvars, wantfields, callback)
+	wg.Wait()
+	// TODO(acharis): possibly better to aggregate errors as in ScatterConn
```

Comment:
that gets pretty tricky if we're going to run both left and right at the same time. if we want the fields, but the left return no rows, then we need the ones from the right.